Implement runtime override option for loadPyodide

Drranny opened the pull request:
**Resolves #5851**

### Description

This PR introduces a runtime override option for `loadPyodide`, enabling users to explicitly specify the runtime environment. The implementation addresses all reviewer feedback from the previous PR.

### Key Changes

* **`runtime` option for `loadPyodide`**
  Users can force runtime detection (`'browser' | 'node' | 'deno' | 'bun'`).
* **`setRuntimeOverride` function**
  Provides programmatic override for testing and development.
* **Improved runtime detection**
  More accurate detection logic for all supported runtimes.
* **Runtime flags moved to `pyodide.ffi`**
  Removed monkey patching on `sys` module.
* **Singleton pattern with `globalThis`**
  Ensures consistent global state across references using `globalThis.__PYODIDE_RUNTIME_ENV__`.
* **Bundle consistency fix**
  Prevents state mismatches between `pyodide.mjs` and `pyodide.asm.wasm` by storing runtime state in `globalThis`.

### Technical Implementation

* **Global state management**: Used `globalThis.__PYODIDE_RUNTIME_ENV__` to maintain a single source of truth across different Pyodide bundles
* **Bundle-safe singleton**: Ensures `RUNTIME_ENV` is shared between `pyodide.mjs` and `pyodide.asm.wasm` without state inconsistencies
* **Clean override mechanism**: `setRuntimeOverride` directly modifies the global runtime environment object

### Reviewer Feedback Addressed

* Implemented singleton with global state consistency
* Removed `sys` monkey patching, moved to `pyodide.ffi`
* Fixed bundle consistency issues using `globalThis`
* Clear separation of runtime detection from main API

### Testing

* 11 unit tests added for runtime override and detection logic.
* Verified across Node.js, Browser, Deno, and Bun environments.

### Checklist

* [ ] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry
* [x] Add / update tests
* [ ] Add / update outdated documentation

Drranny commented:
Changed true → True in runtime flag initialization to fix NameError.

ryanking13 commented on src/js/test/unit/environments.test.ts:
This unittest looks okay. But let's write an integration test too.

I think you can add one new testcase in [test_pyodide.py](https://github.com/pyodide/pyodide/blob/main/src/tests/test_pyodide.py) with the following scenario.

1. Run test in the browser environment, but set some fake global variables that deceive Pyodide into thinking it is running in Node.

2. Call loadPyodide without setting the runtime. It should fail.

3. Call loadPyodide with setting the runtime, it should pass.

Drranny commented:
Hi @ryanking13,

Thank you for the detailed review! I've addressed all the comments you mentioned:

All Issues Addressed:

1. Infinite loop prevention: Modified `updateDerivedFlags()` to take `runtimeEnv` as parameter, breaking the circular dependency with `getGlobalRuntimeEnv()`.

2. Object.keys() for flag reset: Implemented `Object.keys(runtimeEnv).forEach(key => runtimeEnv[key] = false)` to prevent human error in `overrideRuntime()`.

3. Node-compatible runtimes: Added `runtimeEnv.IN_NODE = true` for both Deno and Bun cases since they are Node-compatible.

4. IN\_NODE\_COMMONJS calculation: Added proper calculation logic in `updateDerivedFlags()`.

5. Comment location: Moved the main comment to `getGlobalRuntimeEnv()` function as requested.

6. Integration test: Added the requested test case in `test_pyodide.py` with the exact scenario:

-    Run in browser environment with fake Node.js globals
-    `loadPyodide()` without runtime should fail
-    `loadPyodide({ runtime: "browser" })` should pass

Chrome/Firefox CI/CD Failures:

However, I'm experiencing persistent failures in Chrome/Firefox tests for the new integration tests:

```
FAILED test_pyodide.py::runtime_environment_override_integration[chrome] - AssertionError: loadPyodide should succeed with explicit runtime setting
```
Question: Could this be related to the `test_pyodide.py` implementation, or might there be an issue with how the runtime override mechanism works in the browser environment?

The tests pass locally in Node.js environment, but fail consistently in Chrome/Firefox CI. I've added extensive debugging logs and improved async handling, but the issue persists.

Would you mind taking a look at the test implementation to see if there's something I'm missing? I want to make sure the integration test properly validates the runtime override functionality as intended.

Thank you for taking the time to review this!

Drranny commented:
> > The tests pass locally in Node.js environment, but fail consistently in Chrome/Firefox CI. I've added extensive debugging logs and improved async handling, but the issue persists.
> > Would you mind taking a look at the test implementation to see if there's something I'm missing? I want to make sure the integration test properly validates the runtime override functionality as intended.
> 
> That's a good catch @Drranny. I noticed that updating the Pyodide-side runtime detection is not enough as Emscripten also has it's own environment detection system (sadly...) ([code pointer](https://github.com/emscripten-core/emscripten/blob/a37d6bc29514396a32d681fdb6914d3904e78a41/src/shell.js#L62-L71))
> 
> Unfortunately, I think we need to discuss this with Emscripten people as well, as it looks like they only allow overriding those values in the compile time, not in runtime.

That’s a relief — I was assuming this was an issue with my code and tried several fixes, but it didn’t work out. If this limitation really comes from Emscripten’s environment detection, then I suppose we’ll need to hold off on that part for now.

I’ve also updated the other parts based on your previous review comments. When you have time, could you please check if I’ve addressed them correctly? 🙏

Thanks again for your guidance!

Drranny commented:
> @Drranny Thanks for the understanding. I'll open an issue in Emscripten to find out if we can unblock this task.
> 
> Meanwhile, I think we can split this PR into two PRs.
> 
> 1. Introducing a singleton RuntimeEnv object for environment detection, which is a refactoring PR.
> 2. Implementing `overrideRuntime` for runtime overriding (this one is blocked)
> 
> So, how about making a separate PR for 1? I think you already implemented all parts, so just cherry picking changes in this PR would work.

Sorry for the late follow-up, and thanks again for the detailed review! 🙇‍♀️
Just to confirm before I proceed — would you prefer that I split this PR first (refactoring , runtime override) and then address the review comments in the refactoring PR,
or should I address the comments here first and then split afterward?

ryanking13 commented:
> Just to confirm before I proceed — would you prefer that I split this PR first (refactoring , runtime override) and then address the review comments in the refactoring PR,
or should I address the comments here first and then split afterward?

I think the former: creating a split PR and addressing comment there, would be better for me to review.
